xla/service/gpu/cudnn_norm_rewriter.cc

```diff
@@ -0,0 +1,652 @@
+                                ->literal()
+                                .GetAsDouble({})
+                                .value();
+      int64_t numel = 1;
+      for (auto dim : norm_dims) {
+        numel *= reduce->operand(0)->shape().dimensions()[dim];
+      }
+      float rnumel = 1. / float(numel);
+      float numerical_epsilon = 128 * std::numeric_limits<bfloat16>::epsilon();
+      if (abs(actual_rnumel - rnumel) >
+          (abs(actual_rnumel + rnumel) * numerical_epsilon)) {
+        VLOG(1) << "Expectation scaling factor outside tolerance.";
+        return OkStatus();
+      }
+
+      // If necessary, transpose the input so that the dimensions not being
+      // normalized are the leading dimensions.
+      std::vector<int64_t> non_norm_dims;
+      for (int64_t dim = 0; dim < input->shape().rank(); ++dim) {
+        if (std::find(norm_dims.begin(), norm_dims.end(), dim) ==
+            norm_dims.end()) {
+          non_norm_dims.emplace_back(dim);
+        }
+      }
+      std::vector<int64_t> transpose_order = non_norm_dims;
+      transpose_order.insert(transpose_order.end(), norm_dims.begin(),
+                             norm_dims.end());
+
+      bool apply_transpose = false;
+      for (int i = 0; i < transpose_order.size(); ++i) {
+        if (transpose_order[i] != i) {
+          apply_transpose = true;
+          break;
+        }
+      }
+
+      std::optional<HloInstruction*> transpose;
+      std::vector<int64_t> inverse_transpose_order(transpose_order.size());
+      if (apply_transpose) {
+        for (int k = 0; k < transpose_order.size(); ++k) {
+          inverse_transpose_order[transpose_order[k]] = k;
+        }
+        TF_ASSIGN_OR_RETURN(transpose,
+                            MakeTransposeHlo(input, transpose_order));
+      }
+
+      // Combine the dimensions not normalized into the first dimension of the
+      // input as required by cuDNN.
+      std::vector<int64_t> reshaped_dims = {1};
+      for (auto non_norm_dim : non_norm_dims) {
+        reshaped_dims[0] *= input->shape().dimensions(non_norm_dim);
+      }
+      for (auto norm_dim : norm_dims) {
+        reshaped_dims.emplace_back(input->shape().dimensions(norm_dim));
+      }
+      // cuDNN requires tensors to have at least four dimensions.
+      while (reshaped_dims.size() < 4) {
+        reshaped_dims.emplace_back(1);
+      }
+
+      Shape reshaped_shape =
+          ShapeUtil::MakeShape(input->shape().element_type(), reshaped_dims);
+      TF_ASSIGN_OR_RETURN(
+          HloInstruction * reshape,
+          MakeReshapeHlo(reshaped_shape, transpose.value_or(input)));
+
+      // Reshape the scale and bias.
+      std::vector<int64_t> reshaped_scale_dims(reshaped_dims.begin() + 1,
+                                               reshaped_dims.end());
+      // cuDNN requires tensors to have at least four dimensions.
+      while (reshaped_scale_dims.size() < 4) {
+        reshaped_scale_dims.emplace_back(1);
```

Comment:
Shouldn't the shape of scale in your example be [3, 4, 5] instead of [1, 3, 4, 5]? We only augment it to [3, 4, 5, 1] because cuDNN requires inputs to have at least four dimensions.

Reply:
> Shouldn't the shape of scale in your example be [3, 4, 5] instead of [1, 3, 4, 5]? We only augment it to [3, 4, 5, 1] because cuDNN requires inputs to have at least four dimensions.

The [table in the cudnn documentation for the NormalizationForward engine](https://docs.nvidia.com/deeplearning/cudnn/developer-guide/index.html#normalizationforward) shows the scale/bias should have shape `[1,C,(D),H,W]`:

![image](https://github.com/openxla/xla/assets/6510203/a575fb08-2906-4a62-a847-2fd018992fd0)

So the first dimension is always expected to be 1, according to the documentation.